File: api/schemas/message.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class MessageSendRequest(BaseModel):
    """发送一条私聊消息。"""

    from_user_id: int = Field(..., description="发送方用户 id")
    to_user_id: int = Field(..., description="接收方用户 id")
    content: str = Field(..., min_length=1, description="消息内容")


class MessageReadRequest(BaseModel):
    """标记某用户收到的若干消息为已读。"""

    to_user_id: int = Field(..., description="接收方用户 id")
    ids: list[int] = Field(default_factory=list, description="要标记已读的消息 id 列表")


class UserMessage(BaseModel):
    """一条用户消息。"""

    id: int
    from_user_id: int
    to_user_id: int
    msg_type: str = "chat"
    content: str = ""
    payload: dict[str, Any] | None = None
    is_delivered: bool = False
    delivered_at: datetime | None = None
    is_read: bool = False
    created_at: datetime | None = None

File: front/mini-app/src/api/message.ts
import { request } from '@/utils/request'
import { getStoredUserId } from '@/utils/auth'

export type MessageType = 'chat' | 'egg_match_notify' | string

export interface UserMessage {
  id: number
  from_user_id: number
  to_user_id: number
  msg_type: MessageType
  content: string
  payload: Record<string, any> | null
  is_delivered: boolean
  delivered_at: string | null
  is_read: boolean
  created_at: string | null
}

/** 收件箱：我收到的消息 / 通知（按时间倒序）。 */
export function fetchInbox(userId: number | string, limit = 50) {
  return request<UserMessage[]>({
    url: '/api/messages/inbox',
    data: { user_id: userId, limit },
  })
}

/** 未读消息数（用于红点）。 */
export function fetchUnreadCount(userId: number | string) {
  return request<{ count: number }>({
    url: '/api/messages/unread-count',
    data: { user_id: userId },
  })
}

/** 发送一条私聊消息（REST 兜底，实时走 ws）。 */
export function sendMessage(fromUserId: number | string, toUserId: number | string, content: string) {
  return request<UserMessage>({
    url: '/api/messages',
    method: 'POST',
    data: { from_user_id: fromUserId, to_user_id: toUserId, content },
  })
}

/** 标记若干消息为已读。 */
export function markMessagesRead(toUserId: number | string, ids: number[]) {
  return request<{ ok: boolean }>({
    url: '/api/messages/read',
    method: 'POST',
    data: { to_user_id: toUserId, ids },
  })
}

/** 拉取两个用户间的历史会话。 */
export function fetchConversation(userA: number | string, userB: number | string, limit = 50) {
  return request<UserMessage[]>({
    url: '/api/messages/conversation',
    data: { user_a: userA, user_b: userB, limit },
  })
}

/**
 * 刷新「更多」tab 的红点：未读 > 0 显示，否则隐藏。index 3 = 底部「更多」按钮。
 * 未登录时静默跳过。
 */
export async function refreshMoreTabRedDot(): Promise<number> {
  const userId = getStoredUserId()
  if (!userId) return 0
  try {
    const { count } = await fetchUnreadCount(userId)
    if (count > 0) {
      uni.showTabBarRedDot({ index: 3 })
    } else {
      uni.hideTabBarRedDot({ index: 3 })
    }
    return count
  } catch {
    return 0
  }
}
